core/trino-spi/src/main/java/io/trino/spi/predicate/SortedRangeSet.java

```diff
@@ -79,10 +80,11 @@ public final class SortedRangeSet
 
     private final boolean[] inclusive;
     private final Block sortedRanges;
+    private volatile Optional<Boolean> discreteSet;
 
     private int lazyHash;
 
-    private SortedRangeSet(Type type, boolean[] inclusive, Block sortedRanges)
+    private SortedRangeSet(Type type, boolean[] inclusive, Block sortedRanges, Optional<Boolean> discreteSet)
```

Comment:
What does Optional.empty() mean for discreteSet?

Reply:
> What does Optional.empty() mean for discreteSet?

This is a lazy boolean determining if set is discrete or not.
* If we know set is discrete (either we checked each range or explicit discrete set constructor was used) then this is `true`.
* if we checked each range and the set is not discrete, then this is `false`.
* if we don't know if set is discrete or not, then this is `empty`.

The idea is to avoid explicit expensive discrete set check if not needed.

Reply:
Maybe it would be more readable as an Enum with EMPTY, TRUE, FALSE values ?

Reply:
> Maybe it would be more readable as an Enum with EMPTY, TRUE, FALSE values ?

We could do that, but I'm not sure this will improve readability.

Reply:
Added enum